### ISSUE TITLE
Provide a single executable cell

### ISSUE THREAD

**Issue by benabel** [jupyterlite/jupyterlite#148]
Hi dear jupyterians, 

I'm a french teacher and I really like this project because it's quite difficult for students to install python on their own machines.

First of all the full notebook frontend would be very useful for practicing on projects.

But sometimes we just need a single cell to show a quick example or ask a question. Is there's any component in the jupyterstack that could be ported as something as simple as a web component as it is done on `wc-component`.

https://github.com/vanillawc/wc-code

Thanks for sharing this as an open source project.

**Comment by jtpio** [comment]
Thanks @benabel.

JupyterLab Code Consoles should already work in JupyterLite, and might be a good fit for this use case?

https://user-images.githubusercontent.com/591645/122247036-5b451580-cec7-11eb-8a3a-4ff25a537fcd.mp4

Ideally this would also be supported in the Retro interface (still to be implemented) to have a single console / cell on the page:  https://github.com/jupyterlab/retrolab/issues/49

**Comment by benabel** [comment]
Thanks for your answer,, console seems to do the job if it is alone one the page. 

Is there any way to provide initial content in the console, some lines of code to be executed  and or modified by the user?

**Comment by jtpio** [comment]
> Is there any way to provide initial content in the console, some lines of code to be executed and or modified by the user?

The code console is more of a temporary and stateless component. It sounds like a notebook would be quite appropriate in that case?

**Comment by psychemedia** [comment]
Just as an example of a related user experience, [SageMathCell](https://sagecell.sagemath.org/) allows you to run an arbitrary bit of code against a CoCalc kernel and display the result.

![image](https://user-images.githubusercontent.com/82988/122429023-5a28ec80-cf8a-11eb-9046-32fb3d26fdf0.png)

**Comment by bollwyvl** [comment]
Right. It's a good goal. Once we have #118 squared away, we'll be able to do all kinds of custom apps (without substantially over-inflating the final archive (or having to do multiple releases), with a file tree more like:

```
app/
  index.html    # will redirect to... something
  static/       # where all the build stuff will end up
  lab/         
    index.html  # won't be required
    extensions/
  retro/
    index.html
  editor/
    index.html
  cell/
    index.html
  term/
    index.html
  book/
    index.html  # will need lots of work!
  # .../        # basically anything from https://github.com/jupyterlab/jupyterlab/tree/master/examples
```

As for integration of that `cell`: I anticipate the initial approach being an `iframe` with a (documented) URL param mini-language as the most-broadly integrateable solution. 

```html
<body>
  <h1>run some code</h1>
 <iframe src="./_static/cell/index.html?p=foo.ipynb&d=Notebook&m=io" />
</body>
```

We can emit a few things over `postMessage` (with a schema) but generally won't have access to the outer scope, and _would_ rely on the source documents being in `_static/files`, but _wouldn't_ need them indexed.

Anyhow, maybe something like ![a mockup](https://raw.githubusercontent.com/gist/bollwyvl/a3c6662e8202c67a44024baabf46eb3e/raw/1e3487a21dfbfa4dc7c2e09b821522786a682a94/One%20Cell.dio.svg)

**Comment by benabel** [comment]
This architecture and cell ui seems really promising @bollwyvl.

Maybe a webcomponent could be used with iframe inside. They are well supported now and it could allow a common html attribute syntax.

**Comment by psychemedia** [comment]
In passing, I note https://terminal.mybinder.org/ . Is that the sort of thing that `term/` would equate to?

And is `book/` for Jupyter Book, eg allow cells to run against the wasm kernel via some sort of thebe-like clone?

(Just trying to get my head round where your roadmap is going. This is all v exciting!:-) )

**Comment by psychemedia** [comment]
And another observation: ThebeLab typically launches against a remotely started BinderHub kernel with a js config of the form:

```
    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "ouseful-course-containers/ou-tm112-notebooks",
            ref: "master",
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
```

But the package also supports connections to outstanding servers ([example](https://thebelab.readthedocs.io/en/latest/_static/local.html)) with a setting of the following form and a Jupyter notebook server running with appropriate ` --NotebookApp.token` and `--NotebookApp.allow_origin` settings:


```
<script type="text/x-thebe-config">
{
  bootstrap: true,
  kernelOptions: {
    name: "python3",
    serverSettings: {
      "baseUrl": "http://127.0.0.1:8888",
      "token": "test-secret"
    }
  },
}
</script>
```

I created a related issue here: https://github.com/executablebooks/thebe/issues/412

**Comment by jtpio** [comment]
This PR adds support for code consoles in JupyterLite Retro: https://github.com/jupyterlite/jupyterlite/pull/313

Which can be tested on the RTD preview with the following link:

https://jupyterlite--313.org.readthedocs.build/en/313/_static/retro/consoles/?path=/console-1-acbc7544-d4ae-4fbe-a72b-5a2e15c26e63

This can be a good first step for having a very simple and minimalist cell on a browser page.

> Anyhow, maybe something like ![a mockup](https://raw.githubusercontent.com/gist/bollwyvl/a3c6662e8202c67a44024baabf46eb3e/raw/1e3487a21dfbfa4dc7c2e09b821522786a682a94/One%20Cell.dio.svg)

We should track that idea upstream. Maybe this could be a setting of the JupyterLab code console to toggle between the current layout and the side-by-side view?

**Comment by jtpio** [comment]
> This can be a good first step for having a very simple and minimalist cell on a browser page.

The current JupyterLab code console has the cell at the bottom of the page when rendered in Retro. This is not ideal especially for educational purposes or when demoing something. So the vertical split mode mentioned above would be greatto bring the code editor to the top left part of the page.

**Comment by jtpio** [comment]
> We should track that idea upstream.

This issue can be a good place to have that discussion: https://github.com/jupyterlab/jupyterlab/issues/5293

**Comment by jtpio** [comment]
FYI I have recently been iterating on REPLite (https://github.com/jtpio/replite) to provide a minimal code console that is easy to embed on other websites with:

```html
<iframe src="https://replite.vercel.app/retro/consoles?toolbar=1&kernel=python&code=import numpy as np" width="100%" height="100%">
</iframe>
```

The motivation was to make it easier it to try `numpy` in the browser on [numpy.org](https://numpy.org/): https://github.com/numpy/numpy.org/pull/547

![image](https://user-images.githubusercontent.com/591645/152338112-c2d41c11-a7a8-4d00-9c39-d918a83fe4d5.png)


For now it's just the RetroLab console page (`/retro/consoles`) with custom CSS and plugins to read options for the URL query arguments.

Ideally there should be a more generic way for building small lab and lite remixes from scratch with less effort (https://github.com/jupyterlab/jupyterlab/pull/11693). Or we ship a new `cell` app as mentioned above once #118 is fixed.

**Comment by jtpio** [comment]
One approach for having a single cell / console app in JupyterLite could be to do something like for RetroLab:

- develop the new lab-based app in a separate repo, so it's also usable with stock JupyterLab and regular kernels
- release the JS packages
- consume the JS packages here under `app/console` or `app/cell`

An alternative could be to start looking into allowing for "federated" apps, just like for extensions. So it could be developed separately and shipped on deploy.

The `jupyterlite` package currently defines the list of apps here: https://jupyterlite.readthedocs.io/en/latest/api/py/jupyterlite.html#jupyterlite.constants.JUPYTERLITE_APPS

**Comment by bollwyvl** [comment]
> usable with stock

big fan of re-use... but i think some of these specific, embeddable cases make sense to initially be built and tuned here in core, at least for a while. we'd want to release the js packages _anyway_ such that someone could extend them. likely we'll find some more commonality that we want to keep in the server, etc.

as to re-using e.g. single-cell with a real server: i dunno... i guess i could see the case, if one had, say, a richer outer word processing/drawing editor experience, and wanted to embed some cells in iframes. but just because it _looks_ simple, it would _still_ be a dangerous application if it enables all the endpoints. every time i see someone ask, _I "just" want to run Jupyter in my django app_, and then see stuff about listening to `*` and turning off tokens, i freak out a little bit.

>  allowing for "federated" apps, just like for extensions.

again, hooray "federate everything we possibly can". as long as there's a `pip` or `conda` solver out there making sure that the dependencies can be documented without trying to run the thing in the browser, this will work, and as mentioned elsewhere, we'd ideally be able to bump the upstream lab release and get it on "leaf" apps like cell and console.

**Comment by jtpio** [comment]
What do folks think about having something similar to the `cell` example in the JupyterLab repo? https://github.com/jupyterlab/jupyterlab/tree/master/examples/cell

![image](https://user-images.githubusercontent.com/591645/157723237-61b8d93d-cd77-48a9-b394-565094845695.png)

That way we keep the vertical layout like the current `repl` app, so it renders better on smaller layouts (for example on mobile), but the cell is now located at the top closer to the toolbar.

cc @martinRenou who might be interested

**Comment by jtpio** [comment]
Or we update the upstream JupyterLab Code Console to make it possible to place the input cell at the top?

**Comment by psychemedia** [comment]
I really struggle with the console having the command line way down the bottom of the cell after lots of white space. I'm not sure how the cell at the top would feel as commands are issued and results generated. Maybe the  output would append to end with an automated scroll bar so the output is still near ish the top bar command line cell.

**Comment by bollwyvl** [comment]
> update the upstream JupyterLab Code Console to make it possible to place the input cell at the top?

Yeah, that would be a reasonable setting/command and would likely be solveable with e.g. a single CSS class. 

Meanwhile, a side-by-side console would _also_ be interesting. Similarly, as flexible as it is, I think lab's side-by-side notebook rendering options didn't go _far enough_ to making these configurable (side-by-side, but what about top-to-bottom, etc.)

> cell is now located at the top closer to the toolbar.

Anything we pick and set in stone will be wrong for one use case or another :crying_cat_face: 

I think for this app, "one cell" could really be a data-driven confection of any number of cell parts (i thought about this _a lot_ back in the day on `nbpresent`). This whole thing could be in an `overrides.json`, which would be _relatively_ easy to iterate on in a "full" lab lite.

One could have:
- just the output
  - this would hit _really hard_ with a single layout ipywidget stuffed with crazy things
- an output and a toolbar with run-all-cells
  - good for e.g. reloading a graph
- toggleable parts
- the inspector ne contextual help

**Comment by jtpio** [comment]
> Or we update the upstream JupyterLab Code Console to make it possible to place the input cell at the top?

This seems to be tracked upstream in https://github.com/jupyterlab/jupyterlab/issues/1058

**Comment by joelostblom** [comment]
I'm also interested in this. There is a nice implementation for what the interface could look like at https://quarto.thecoatlessprofessor.com/pyodide/

![image](https://github.com/jupyterlite/jupyterlite/assets/4560057/5b96d44f-6671-4f5e-a7e2-d2afeda8679e)

Which is the same as in the R version of that addon https://quarto-webr.thecoatlessprofessor.com/. Especially for running code as part of interactive exercises in a text book (https://github.com/jupyterlite/jupyterlite/issues/1333), I think it would be really helpful to have an explicit "Run" button (and maybe also an optional "Submit" button)

**Comment by jtpio** [comment]
Thanks @joelostblom for sharing :+1: 

Yes having a single cell is still very much doable. And as mentioned above it could be developed as a regular JupyterLab extension, so it could also be used in JupyterLab or other lab-based applications.

**Comment by jtpio** [comment]
This is now coming soon to JupyterLite via https://github.com/jupyterlite/jupyterlite/pull/1570:

![Image](https://github.com/user-attachments/assets/15002f39-8b3b-46a6-80be-e66e02328962)

Some follow-ups to https://github.com/jupyterlite/jupyterlite/pull/1570 would be:

- document how to use the new code console options (prompt position, clear cells, show banner...)
- provide an example `overrides.json` to set some of these options
- expose the new options via query parameters
- think about what other options could be useful as part of the REPL app (copy code, more toolbar buttons?)

**Comment by agriyakhetarpal** [comment]
Since you've asked, a physical "send" button that can be pressed like in instant messaging apps next to the code snippet would be quite nice. It would allow the use of a button to input code into the REPL, as an alternative to <kbd>shift</kbd>+<kbd>Enter</kbd>. This would be useful for mobile users, where one has to use the terminal mode setting at the moment as the shift key isn't available.

Alternatively, a "terminal-mode-on-mobile" could be provided as well, but that's only half as neat a solution.

**Comment by agriyakhetarpal** [comment]
> think about what other options could be useful as part of the REPL app (copy code, more toolbar buttons?)

@jtpio, since you've asked, a physical "send" button that can be pressed like in instant messaging apps next to the code snippet would be quite nice. It would allow the use of a button to input code into the REPL, as an alternative to <kbd>shift</kbd>+<kbd>Enter</kbd>. This would be useful for mobile users, where one has to use the terminal mode setting at the moment as the shift key isn't available.

Alternatively, a "terminal-mode-on-mobile" could be provided as well, but that's only half as neat a solution.

**Comment by psychemedia** [comment]
I don't the repl does close this issue does it? It doesn't offer a single executsble cell? It offers a repl.. if you enter code in the cell and run it you get a second cell? Or can you disable that from appearing? And also reset the state of the single executsble cell back to a clean slate before you run code in it again?

**Comment by jtpio** [comment]
> if you enter code in the cell and run it you get a second cell?

If you try the link posted just above, which uses `clearCodeContentOnExecute=0`, the content stays in the cell after execute.

> And also reset the state of the single executsble cell back to a clean slate before you run code in it again?

There is still a kernel behind, so hidden state for sure. If you think there should be an option to automatically restart the kernel on execute, please open a new issue, thanks!

**Comment by psychemedia** [comment]
Ah, thanks.. not sure what demo settings I'd clixked through on elsewhere...
